### PR TITLE
Detect HAVE_LARGE_STACKS at compile time

### PR DESCRIPTION
**Pull request by kev009**
Signed-off-by: Kevin Bowling <kbowling@FreeBSD.org>

### Motivation and Context
On FreeBSD, this OZFS change seems to regress performance of sequential workloads (find /, indexing a mail spool, etc) handily versus the previous implementation.  I am unsure if the conditions of b58986eebf3c47c946393da4b968ee33edaea99e apply equally to FreeBSD stack usage, but we have 16K stacks on common platforms so detect and set HAVE_LARGE_STACKS when it is the case.

Discussed with: @mjguzik 

CC @allanjude @amotin 

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by allanjude — APPROVED** [review]
Reviewed By: Allan Jude <allanjude@freebsd.org>

**Comment by allanjude** [comment]
Thanks for looking into this. Do you happen to have any numbers to show how much of a difference this makes?

File: config/kernel-config-defined.m4 [hunk 1]
```diff
@@ -55,6 +55,25 @@ AC_DEFUN([ZFS_AC_KERNEL_SRC_CONFIG_THREAD_SIZE], [
...
+dnl # Check configured KSTACK_PAGES * pagesize
+dnl #
+dnl # On FreeBSD we check these runtime settings, it's only expected to be false
+dnl # on 32b mips and arm with default configs.
+dnl #
+AC_DEFUN([ZFS_AC_KERNEL_CONFIG_THREAD_SIZE_FREEBSD], [
+    AC_MSG_CHECKING([whether kernel was built with 16K or larger stacks])
+	PAGE_SIZE=`sysctl -n hw.pagesize`
+	KSTACK_PAGES=`sysctl -n kern.kstack_pages`
+	KSTACK_SIZE=`expr $KSTACK_PAGES \* $PAGE_SIZE`
```
Comment (amotin) [inline config/kernel-config-defined.m4:68]: Thank you for catching this, but you are checking variables of the running system, not a system you are building for, better represented by kernel headers.  For the mentioned mips and arm systems cross-building on x86 is quite popular, so your patch will be wrong. You should probably include sys/param.h and take KSTACK_PAGES and PAGE_SIZE from there.

Comment (kev009) [reply config/kernel-config-defined.m4]: @amotin thanks have a look at this update

File: config/kernel-config-defined.m4 [hunk 2]
```diff
@@ -55,6 +55,29 @@ AC_DEFUN([ZFS_AC_KERNEL_SRC_CONFIG_THREAD_SIZE], [
...
+dnl # on 32b mips and arm with default configs.
+dnl #
+AC_DEFUN([ZFS_AC_KERNEL_CONFIG_THREAD_SIZE_FREEBSD], [
+	AC_MSG_CHECKING([whether kernel was built with 16K or larger stacks])
+	AC_COMPILE_IFELSE([
+		AC_LANG_SOURCE([
+			#include <sys/param.h>
+			#if (KSTACK_PAGES * PAGE_SIZE < 16384)
+			#error "THREAD_SIZE is less than 16K"
+			#endif
```
Comment (amotin) [inline config/kernel-config-defined.m4:71]: This part looks better to me now.

File: module/Makefile.bsd [hunk 3]
```diff
@@ -68,6 +68,10 @@ CFLAGS+= -DBITS_PER_LONG=32
 CFLAGS+= -DBITS_PER_LONG=64
 .endif
 
+.if ${MACHINE_CPUARCH} != "mips" && ${MACHINE_ARCH} != "arm"
+CFLAGS+= -DHAVE_LARGE_STACKS=1
+.endif
+
```
Comment (amotin) [inline module/Makefile.bsd:74]: But here I am at loss: I haven't looked on the build process, but it looks weird to me first detect the stack size in configure, but then hardcode it.

Comment (kev009) [reply module/Makefile.bsd:74]: This is a non-autotools build and is self-sufficient BSDmakefile.  It is done for completeness.

Comment (behlendorf) [reply module/Makefile.bsd:74]: Just to confirm i386 and powerpc use 16k stacks by default as well?

I'm also not sure the autoconf change really buys us much if we're just going to hardcode it here, but I'm not opposed to it.

Comment (kev009) [reply module/Makefile.bsd:74]: Yes, both of those use 4 stack pages on FreeBSD.

There seems to be a misunderstanding of what this file even is, which is orthogonal to this changeset.  It's a self-contained BSDmakefile to build the kmod and does not tie in to autotools at all, it uses the bsd.mk framework.  Whether it is unnecessary cruft maybe is something you are both encircling.

Comment (behlendorf) [reply module/Makefile.bsd:74]: What I was driving at is the `Makefile.bsd` file is used by the normal `make modules` target in this repository as well.  So there shouldn't be a need to add it to the autoconf bits.

Comment (kev009) [reply module/Makefile.bsd]: Ok which side should I remove it from?

Comment (behlendorf) [reply module/Makefile.bsd:74]: That looks better to me and I believe it should be sufficient.  Thanks.

Comment (kev009) [reply module/Makefile.bsd:74]: Ok I see the tie-in now and this will more closely match the downstreaming.

**Comment by amotin** [comment]
Thinking about it again, why would it need to be done at configure or even make?  If we already have the defines, why not just check them in some header(s)?  Why complicate?

**Comment by kev009** [comment]
Feel free to do it another way, I don't know this code very well.

**Comment by amotin** [comment]
I don't mean anything complicated.  Not sure if there is a better header, may be @behlendorf propose one, but I am thinking about something like this, plus equivalent for Linux:
```
diff --git a/sys/contrib/openzfs/include/os/freebsd/zfs/sys/zfs_context_os.h b/sys/contrib/openzfs/include/os/freebsd/zfs/sys/zfs_context_os.h
index 8dbe907d098..a32eb52c53c 100644
--- a/sys/contrib/openzfs/include/os/freebsd/zfs/sys/zfs_context_os.h
+++ b/sys/contrib/openzfs/include/os/freebsd/zfs/sys/zfs_context_os.h
@@ -41,6 +41,10 @@
 #include <sys/ccompat.h>
 #include <linux/types.h>
 
+#if KSTACK_PAGES * PAGE_SIZE >= 16384
+#define        HAVE_LARGE_STACKS       1
+#endif
+
 #define        cond_resched()          kern_yield(PRI_USER)
 
 #define        taskq_create_sysdc(a, b, d, e, p, dc, f) \
diff --git a/sys/contrib/openzfs/lib/libspl/include/os/freebsd/sys/zfs_context_os.h b/sys/contrib/openzfs/lib/libspl/include/os/freebsd/sys/zfs_context_os.h
index f5a136d2212..0e9cfd77ddc 100644
--- a/sys/contrib/openzfs/lib/libspl/include/os/freebsd/sys/zfs_context_os.h
+++ b/sys/contrib/openzfs/lib/libspl/include/os/freebsd/sys/zfs_context_os.h
@@ -30,5 +30,6 @@
 #define        ZFS_CONTEXT_OS_H_
 
 #define        ZFS_EXPORTS_PATH        "/etc/zfs/exports"
+#define        HAVE_LARGE_STACKS       1
 
 #endif
```

**Comment by behlendorf** [comment]
I think either approach should work fine, although doing the check as `KSTACK_PAGES * PAGE_SIZE >= 16384` on FreeBSD seems like it would be more robust.  On Linux we should be able to move that check out of autoconf and in to the `openzfs/include/os/linux/zfs/sys/zfs_context_os.h` header.  Note that in user space we already always set `HAVE_LARGE_STACKS=1` in `config/Rules.am`.

**Review by amotin — APPROVED** [review]

**Review by freqlabs — APPROVED** [review]
Great! Just the checkstyle whitespace nits to fix :)

**Review by behlendorf — APPROVED** [review]

**Comment by kev009** [comment]
I'd also like this merged back to 2.1, is there anything I can do to help with that?

**Pull request merged**